xml/value.go

```diff
@@ -0,0 +1,329 @@
+package xml
+
+import (
+	"encoding/base64"
+	"fmt"
+	"math"
+	"math/big"
+	"strconv"
+)
+
+// Value represents an XML Value type
+// XML Value types: Object, Array, Map, String, Number, Boolean.
+type Value struct {
+	w       writer
+	scratch *[]byte
+
+	// xml start element is the associated start element for the Value
+	startElement StartElement
+
+	// indicates if the Value represents a flattened shape
+	isFlattened bool
+}
+
+// newValue returns a new Value encoder. newValue does NOT write the start element tag
+func newValue(w writer, scratch *[]byte, startElement StartElement) Value {
+	return Value{
+		w:            w,
+		scratch:      scratch,
+		startElement: startElement,
+	}
+}
+
+// newWrappedValue writes the start element xml tag and returns a Value
+func newWrappedValue(w writer, scratch *[]byte, startElement StartElement) Value {
+	writeStartElement(w, startElement)
+	return Value{w: w, scratch: scratch, startElement: startElement}
+}
```

Comment:
Whats the difference between these two? It looks like `newWrappedValue` will wrap the value within another tag, but the `startElement` is passed further down.

Its kind of confusing when one would be used over the other.  `newValue` seems to be for writing values that are within some other outside defined element., Whereas `newWrappedValue` is for writing values embedded within a provided element?

It looks like `Close` is always called for each type. Wouldn't that mean in the `newValue` case that the value would be closing its parent element that is defined outside of the context of the value?  Maybe some documentation here describing this behavior would help clarify these expectations/behavior more.

Reply:
Renamed the functions to better clarify the purpose.